### ISSUE TITLE
Slight redesign?

### ISSUE THREAD

**Issue by joshday** [JuliaML/PenaltyFunctions.jl#10]
@Evizero I was thinking about how penalties are extremely similar to a scaled loss.  I've been toying around with adding a slightly lower-level interface, something like

```julia
immutable L1Penalty <: ElementPenalty end
value(p::L1Penalty, θ::Number) = abs(θ)
deriv(p::L1Penalty, θ::Number) = sign(θ)
prox{T <: Number}(p::L1Penalty, θ::T, s::T) = soft_thresh(θ, s)
```

I would actually prefer to use this design for SparseRegression.

Then `scaled(L1Penalty(), .1)` would provide back the functionality PenaltyFunctions currently has.  We could put `scaled` (or something else) in LearnBase so LossFunctions/PenaltyFunctions could share. 

Thoughts?  Comments? Suggestions?

**Comment by Evizero** [comment]
I like it!

What would that mean for penalties that have multiple parameters?

**Comment by joshday** [comment]
They would be just like losses that have parameters.  Here's ElasticNet, for example.

```julia
immutable ElasticNetPenalty{T <: Number} <: ElementPenalty 
    α::T 
end
ElasticNetPenalty(α::Number) = (@assert 0 <= α <= 1; ElasticNetPenalty(α))

function value{T <: Number}(p::ElasticNetPenalty{T}, θ::T)
    p.α * value(L1Penalty(), θ) + (1 - p.α) * value(L2Penalty(), θ)
end

function deriv{T <: Number}(p::ElasticNetPenalty{T}, θ::T)
    p.α * deriv(L1Penalty(), θ) + (1 - p.α) * deriv(L2Penalty(), θ)
end

function prox{T <: Number}(p::ElasticNetPenalty{T}, θ::T, λ::T)
    αλ = p.α * λ
    soft_thresh(θ, αλ) / (one(T) + λ - αλ)
end
```

**Comment by Evizero** [comment]
That makes a lot of sense, yea. I am all for it. One thing to be careful about is with assertions/conditions in constructors as they can cause  a pretty noticeable overhead. In LossFunctions I noticed some cases where the overhead was massive. Although that involved checking type parameters.

If the objects are created on the fly (like `L1Penalty` in the elastic net example snippet) then it is probably best to make sure the constructors of those types are as simple as possible. This will shift the responsibility of sensible parameters to the user, but that seems ok for performance sensitive code

**Comment by Evizero** [comment]
I have been thinking about this a little. One thing that is quite annoying is that `typeof(scaled(L1Penalty(),0.1)) <: L1Penalty` is `false`, which is unfortunate.

A question that is maybe a bit out there: Does the scaling factor `lambda` even need to be known in the context of this package? a user could simply do `lambda * value(L1Penalty(), theta)`. Furthermore this would be quite nice for support vector machines, where the unfortunate convention is to scale the loss and not the penalty, i.e. `C * value(loss, y, h(x,theta)) + value(L2Penalty(), theta)`

**Comment by Evizero** [comment]
In the end it is about computing a weighted sum of two building blocks, the loss and the penalty. Maybe the weighing of the terms would better fit into the realm of a "StructuralRisk" rather than any of the two building blocks themselves

**Comment by joshday** [comment]
Are you okay with blowing away master with what's in the josh branch?

**Comment by Evizero** [comment]
You mean merging the `josh` branch into `master`? sure. I am coding in a branch that is based on the `josh` branch currently

**Comment by joshday** [comment]
Cool.  While merge is the correct term, it's going to overwrite everything.  Just checking.

**Comment by joshday** [comment]
Sorry, I didn't see this

> A question that is maybe a bit out there: Does the scaling factor lambda even need to be known in the context of this package? a user could simply do lambda * value(L1Penalty(), theta). Furthermore this would be quite nice for support vector machines, where the unfortunate convention is to scale the loss and not the penalty, i.e. C * value(loss, y, h(x,theta)) + value(L2Penalty(), theta)

That's a fair question.  My goal was to provide both ways (with or without lambda) of using a penalty.  I think I'll mainly be sticking with the without-lambda api.  

> Maybe the weighing of the terms would better fit into the realm of a "StructuralRisk"

I'm definitely open to having scaled losses live somewhere else

**Comment by Evizero** [comment]
Ok yes, fair enough. Offering both makes sense. 

As a side note. The main reason for scaled losses to exist was not for weighting its influence on the risk, but because I think many people are used to for example the L2DistLoss having this 0.5 factor baked in. So in that sense it made sense to me to be able to create a new entity that offers this. So in that sense scaling a loss is really intended to create a new type of loss on the same level as any other loss. It also makes sense if one just wants to use the loss itself without any risk or other thing, but I guess that also applies to Penalties.

Bottom line I think we should somehow ensure that a user does **not**  create a scaled loss or penalty for the sake of regularisation (i.e. setting the hyper-parameter `lambda`). This would be bad! In the case of linear SVMs for example I want to dispatch on the regularisation being a `L1Penalty` or `L2Penalty`, or the loss being `L1HingeLoss` or `L2HingeLoss`, regardless of the regularisation coefficient `C` (or `lambda` for that matter) 

edit: to be clear. it would be bad in the sense that a scaled loss or scaled penalty is a new distinct type and not subtype of the original loss or penalty.

**Comment by joshday** [comment]
How about removing lambda as the parameter?  Something like

```julia
immutable ScaledElementPenalty{P <: ElementPenalty, T <: Number} <: ElementPenalty
    penalty::P
    lambda::T
end
```

It wouldn't look as nice, but then you could dispatch on this, right?
```julia
Union{L1Penalty, ScaledElementPenalty{L1Penalty}}
```

**Comment by joshday** [comment]
(Well, I suppose you could already dispatch on that Union type as is)

**Comment by Evizero** [comment]
Code beauty aside, the one issue i see with that approach is that some function that itself applies some scale factor lambda would in the end scale the original penalty twice. so there would be two points to configure lambda, the loss and then again the function.

In my mind a type-scaled version of a loss or penalty is for when one cares about its value. For example taking losses as inspiration one could define the L1Penalty and L2Penalty one the same level as special cases of the LPPenalty with P=1 and P=2, which is defined as `abs(x)^P`, then one could do type-aliases for those who care about these scaling factors, for example
`typealias RidgePenalty ScaledElementPenalty{L2Penalty, 0.5}`. This would then be a new distinct penalty. 

Well that was my intention with Scaled Losses anyway. Maybe having scaled types does more harm than good

**Comment by Evizero** [comment]
That said, on some level it is convenient to be able to contain all the information into the object and not have to carry around multiple parameters.

**Comment by Evizero** [comment]
What annoys me about my side of the argument is that any function that cares about the penalty being a `L1Penalty` should have no intrinsic problem dealing with a scaled version of it. So given this aspect, your union proposal makes more sense

We could consider having the packages generate typealiases for all the exported losses and penalties in the form of `typealias AbstractMyLoss Union{MyLoss,ScaledLoss{MyLoss}}`. Shouldn't be difficult. these would then serve for dispatching

**Comment by joshday** [comment]
The more I think on this, I think we should just keep scaled Penalties undocumented for now.  I don't yet know how they should be used alongside the unscaled versions, so I don't how they should work.  When one of us runs into something where we want a scaled version, we'll have a better idea of how it should work.

**Comment by Evizero** [comment]
agreed
